File: AlgorithmsPackPythonSolutions/775. Global and Local Inversions.py
/*
This TLEs on LeetCode...
*/

class Solution:
    def isIdealPermutation(self, nums: List[int]) -> bool:
        tmp = nums.copy()

        def merge_sorted_subarrays(nums, start, mid, end):
            global_conversions = 0
            i, j, k = start, mid + 1, start
            while k <= end:
                if i > mid:
                    tmp[k] = nums[j]
                    j += 1
                elif j > end:
                    tmp[k] = nums[i]
                    i += 1
                elif nums[i] < nums[j]:
                    tmp[k] = nums[i]
                    i += 1
                else:
                    tmp[k] = nums[j]
                    j += 1
                    global_conversions += mid - i + 1
                k += 1
            for i in range(start, end + 1):
                nums[i] = tmp[i]
            return global_conversions

        def merge_sort(nums, start, end):
            if start == end:
                return 0
            mid = start + (end - start) // 2
            conv = merge_sort(nums, start, mid)
            conv += merge_sort(nums, mid + 1, end)
            conv += merge_sorted_subarrays(nums, start, mid, end)
            return conv

        n = len(nums)
        tmp = nums.copy()

        local_inversions = sum(nums[i] > nums[i + 1] for i in range(n - 1))
        global_inversions = merge_sort(nums, 0, n - 1)
        return global_inversions == local_inversions

File: AlgorithmsPackJavaSolutions/775. Global and Local Inversions.java
class Solution {
    private int[] tmp;

    private long mergeSortedSubarrays(int[] nums, int start, int mid, int end) {
        long globalConversions = 0;
        for (int i = start, j = mid + 1, k = start; k <= end; k++) {
            if (i > mid)
                tmp[k] = nums[j++];
            else if (j > end)
                tmp[k] = nums[i++];
            else if (nums[i] < nums[j])
                tmp[k] = nums[i++];
            else {
                tmp[k] = nums[j++];
                globalConversions += mid - i + 1;
            }
        }
        System.arraycopy(tmp, start, nums, start, end - start + 1);
        return globalConversions;
    }

    private long mergeSort(int[] nums, int start, int end) {
        if (start == end)
            return 0;
        int mid = start + (end - start) / 2;
        long conv = mergeSort(nums, start, mid);
        conv += mergeSort(nums, mid + 1, end);
        conv += mergeSortedSubarrays(nums, start, mid, end);
        return conv;
    }

    public boolean isIdealPermutation(int[] nums) {
        int n = nums.length;
        tmp = new int[n];

        long localInversions = 0;
        for (int i = 0; i < n - 1; i++) {
            if (nums[i] > nums[i + 1])
                localInversions++;
        }

        long globalInversions = mergeSort(nums, 0, n - 1);
        return globalInversions == localInversions;
    }
}
